src/basic/MAT_FUSED_MUL_ADD-Hip.cpp

```diff
@@ -0,0 +1,308 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "MAT_FUSED_MUL_ADD.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_HIP)
+
+#include "common/HipDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace basic {
+
+#define MAT_FUSED_MUL_ADD_DATA_SETUP_HIP           \
+  const Index_type N = m_N;                        \
+  allocAndInitHipDeviceData(A, m_A, N);            \
+  allocAndInitHipDeviceData(B, m_B, N);            \
+  allocAndInitHipDeviceData(D, m_D, N);			   
+
+#define MAT_FUSED_MUL_ADD_DATA_TEARDOWN_HIP        \
+  getHipDeviceData(m_A, A, N);                     \
+  getHipDeviceData(m_B, B, N);                     \
+  getHipDeviceData(m_D, D, N);                     \
+  deallocHipDeviceData(A);                         \
+  deallocHipDeviceData(B);                         \
+  deallocHipDeviceData(D);						   
+
+__global__ void mat_fused_mul_add_builtin(const Real_ptr A, const Real_ptr B, Real_ptr D, const Index_type N){ 
+  constexpr Index_type Ne = 16;
+  for(Index_type ii = 0; ii != (N/(Ne*Ne)); ++ii){
+  using real4 = __attribute__((__vector_size__(4 * sizeof(Real_type)))) Real_type;
+  real4 result = {0};
+
+  Index_type a_idx = Ne * threadIdx.x + threadIdx.y + ii*(Ne*Ne);
+  Index_type b_idx = threadIdx.x + Ne * threadIdx.y + ii*(Ne*Ne);
+
+  for(Index_type i = 0; i < 4; ++i){
+    Real_type a = A[a_idx];
+    Real_type b = B[b_idx];
+
+#ifdef __gfx90a__	
+#if defined(RP_USE_DOUBLE)
+    result = __builtin_amdgcn_mfma_f64_16x16x4f64(a, b, result, 0, 0, 0);
+#elif defined(RP_USE_FLOAT)
+    result = __builtin_amdgcn_mfma_f32_16x16x4f32(a, b, result, 0, 0, 0);
+#endif  
+#endif
+
+#ifdef __gfx908__
+#if defined(RP_USE_FLOAT)
+    result = __builtin_amdgcn_mfma_f32_16x16x4f32(a, b, result, 0, 0, 0);
+#endif  
+#endif
+    a_idx += 4; // move two columns to the right
+    b_idx += 4*Ne; // move two rows down
+  }
+
+  #pragma unroll 4
```

Comment:
We have RAJA macros for unrolling with a fixed amount: 

https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L572